test/test_cubic_b_spline.cpp

```diff
@@ -0,0 +1,217 @@
+    BOOST_CHECK_SMALL(boost::math::b3_spline_prime<Real>(2), (Real) 0);
+    BOOST_CHECK_SMALL(boost::math::b3_spline_prime<Real>(-2), (Real) 0);
+
+    // Special values:
+    BOOST_CHECK_CLOSE(boost::math::b3_spline<Real>(-1), third<Real>()*half<Real>(), eps);
+    BOOST_CHECK_CLOSE(boost::math::b3_spline<Real>( 1), third<Real>()*half<Real>(), eps);
+    BOOST_CHECK_CLOSE(boost::math::b3_spline<Real>(0), 2*third<Real>(), eps);
+
+    BOOST_CHECK_CLOSE(boost::math::b3_spline_prime<Real>(-1), half<Real>(), eps);
+    BOOST_CHECK_CLOSE(boost::math::b3_spline_prime<Real>( 1), -half<Real>(), eps);
+    BOOST_CHECK_SMALL(boost::math::b3_spline_prime<Real>(0), eps);
+
+    // Properties: B3 is an even function, B3' is an odd function.
+    for (size_t i = 1; i < 200; ++i)
+    {
+        auto arg = i*0.01;
+        BOOST_CHECK_CLOSE(boost::math::b3_spline<Real>(arg), boost::math::b3_spline<Real>(arg), eps);
+        BOOST_CHECK_CLOSE(boost::math::b3_spline_prime<Real>(-arg), -boost::math::b3_spline_prime<Real>(arg), eps);
+    }
+
+}
+/*
+ * This test ensures that the interpolant s(x_j) = f(x_j) at all grid points.
+ */
+template<class Real>
+void test_interpolation_condition()
+{
+    std::cout << "Testing interpolation condition for cubic b splines on type " << boost::typeindex::type_id<Real>().pretty_name()  << "\n";
+    std::random_device rd;
+    std::mt19937 gen(rd());
+    boost::random::uniform_real_distribution<Real> dis(1, 10);
+    boost::random::uniform_real_distribution<Real> step_distribution(0.001, 0.01);
+    std::vector<Real> v(5000);
+    for (size_t i = 0; i < v.size(); ++i)
+    {
+        v[i] = dis(gen);
+    }
+
+    Real step = step_distribution(gen);
+    Real a = dis(gen);
+    boost::math::cubic_b_spline<Real> spline(v.data(), v.size(), a, step);
+
+    for (size_t i = 0; i < v.size(); ++i)
+    {
+        auto y = spline.interpolate_at(i*step + a);
+        // This seems like a very large tolerance, but I don't know of any other interpolators
+        // that will be able to do much better on random data.
+        BOOST_CHECK_CLOSE(y, v[i], 10000000*std::numeric_limits<Real>::epsilon());
+    }
+
+}
+
+template<class Real>
+void test_constant_function()
+{
+    std::cout << "Testing that constants are interpolated correctly by cubic b splines on type " << boost::typeindex::type_id<Real>().pretty_name() << "\n";
+    std::random_device rd;
+    std::mt19937 gen(rd());
+    boost::random::uniform_real_distribution<Real> dis(-100, 100);
+    boost::random::uniform_real_distribution<Real> step_distribution(0.0001, 0.1);
+    std::vector<Real> v(500);
+    Real constant = dis(gen);
+    for (size_t i = 0; i < v.size(); ++i)
+    {
+        v[i] = constant;
+    }
+
+    Real step = step_distribution(gen);
+    Real a = dis(gen);
+    boost::math::cubic_b_spline<Real> spline(v.data(), v.size(), a, step);
+
+    for (size_t i = 0; i < v.size(); ++i)
+    {
+        auto y = spline.interpolate_at(i*step + a);
```

Comment:
We can also test in-between the control points since we know what the value is there?